### ISSUE TITLE
[$1000] IOU - Preview IOU disappears at the end of the conversation after requesting it

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#30147]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.88.3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Open New Expensify app
2. Log in with an expensifail account
3. Navigate to or create a conversation with any participant
4. In the conversation, send an IOU to the user
5. Send multiple messages of any kind to the same user
6. Re-send an IOU to the same user in a conversation.

## Expected Result:
The IOU preview should not disappear from its place. There should be an item in the correct place in the chat history for each IOU. The last such item should have a more expanded IOU preview component with a pay button, showing the total debt from all IOU requests on the report.

## Actual Result:
The IOU preview disappears at the end of the conversation after a request and reappears in its old place in the conversation history

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/b8536421-e172-47de-ac4c-3194340dc7a6

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ee8a739c4993763c</li>
        <li>Upwork Job ID: 1716105961384960000</li>
        <li>Last Price Increase: 2023-12-04</li>
    </ul>
</details>

**Comment by cooldev900** [comment]
I am not reproducible on staging. Is it in production?

**Comment by c3024** [comment]
Is this a bug? All IOUs are merged and made into a single IOU preview. Looks like expected behaviour.

**Comment by goldendragon430** [comment]
Hi, I checked this bug but I didn't find issues like video.
I think this is no problem,

**Comment by goldendragon430** [comment]
Contributor details
Your Expensify account email: corpitts@outlook.com
Upwork Profile Link:  https://www.upwork.com/freelancers/~01598019cf72f02b90

**Comment by goldendragon430** [comment]
Hi, I checked this bug but I didn't find issues like video.
I think this is no problem,

**Comment by andriishupta** [comment]
Hello

I also couldn't reproduce it. Even tried with "Slow 3G" / "Offline mode" - it works without issues

**Comment by andriishupta** [comment]
Contributor details
Your Expensify account email: andrii.shupta+expensify@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~019cf579f0a0b7082a

**Comment by paultsimura** [comment]
May be related to https://github.com/Expensify/App/issues/29607#issuecomment-1771787412

**Comment by Christinadobrzyn** [comment]
asking in https://github.com/Expensify/App/issues/29607 if this will be fixed with that job

**Comment by roryabraham** [comment]
> Is this a bug? All IOUs are merged and made into a single IOU preview. Looks like expected behaviour.

I think we should have only a single IOU preview with a pay button but I think there should be an IOU message in the correct place in the chat history when the original amount was requested. So the video looks not quite right

**Comment by roryabraham** [comment]
> So basically it sounds like we want the IOUs to always be anchored at the bottom of the chat thread

not exactly. There should be an item in the correct place in the chat history for each IOU. The last such item should have a more expanded IOU preview component with a pay button, showing the total debt from all IOU requests on the report

**Comment by Christinadobrzyn** [comment]
Thanks @roryabraham! We'll keep this open and are looking for proposals!

**Comment by Christinadobrzyn** [comment]
Good idea! Upped the price to get more proposals

**Comment by rojiphil** [comment]
@roryabraham
Not sure if I understand the issue. 
Let’s say, 
1) User A generated a money request of amount (say 22) to User B. Report preview gets shown for this.
2) User leaves a normal message to User B
3) User A again generates a new money request (say of amount 8) to User B. Now, a consolidated report preview (with amount 30) will be shown as the last visible action in chat. But then, the report preview of step (1) disappears from it's place.

Are you saying that the money request of step (1) should also be shown in it’s original place in chat?

**Comment by Christinadobrzyn** [comment]
I'm not 100% sure on what we want this to look like @roryabraham would you be able to clarify https://github.com/Expensify/App/issues/30147#issuecomment-1784776721

**Comment by cooldev900** [comment]
## Proposal
from: @cooldev900
### Please re-state the problem that we are trying to solve in this issue.
There should be an item in the correct place in the chat history for each IOU. The last such item should have a more expanded IOU preview component with a pay button, showing the total debt from all IOU requests on the report
### What is the root cause of that problem?
The root cause is that iouAction is not added to REPORT_ACTIONS but reportPreviewAction is only added when we create manual money request so the report page only shows reportPreview chat showing the total debt of all IOU requests in the report.
https://github.com/Expensify/App/blob/9d79fd12ce8f5d1dcb8870b61a8bc4513f55c935/src/libs/actions/IOU.js#L188-L193
### What changes do you think we should make in order to solve the problem?
We have to add iouAction to REPORT_ACTIONS.
```
{
            onyxMethod: isNewChatReport ? Onyx.METHOD.SET : Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${chatReport.reportID}`,
            value: {
                ...(isNewChatReport ? {[chatCreatedAction.reportActionID]: chatCreatedAction} : {}),
                [iouAction.reportActionID]: iouAction,
                [reportPreviewAction.reportActionID]: reportPreviewAction,
            },
 },
```
We can find last iouAction and can eliminate that from sortedAction in here.

https://github.com/Expensify/App/blob/11a7170f19c87f417de84a68fb3815fdbccab20a/src/libs/ReportActionsUtils.ts#L144-L144

But I think it is clear to show all iouActions in Report page so that users can see full history.

After logging out and rejoinning, there are no iouActions in REPORT_ACTIONS. It looks like the backend is removing iouActions from REPORT_ACTIONS, so we need to update the backend to not remove iouActions when sending API calls, I think.

### What alternative solutions did you explore? (Optional)
  N/A

**Comment by Christinadobrzyn** [comment]
@allroundexperts can you review the proposals and let me know if any will work?

**Comment by Christinadobrzyn** [comment]
DMing @allroundexperts for a review

**Comment by roryabraham** [comment]
@allroundexperts your understanding is correct to the best of my knowledge. cc @luacmartins @JmillsExpensify to buddy-check

**Comment by luacmartins** [comment]
correct, it seems like the report previews are duplicated

**Comment by cooldev900** [comment]
@allroundexperts @luacmartins @JmillsExpensify 
I updated [my proposal](https://github.com/Expensify/App/issues/30147#issuecomment-1791407089) to hide last iouAction.

**Comment by Christinadobrzyn** [comment]
Awesome! @allroundexperts would you be able to take a peek at the updated proposal? Thank you!

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We are attempting to implement the solution arrived at [here](https://github.com/Expensify/App/issues/30147#issuecomment-1799529847) i.e.

> First $10 iOU request -> Single preview of $10 shown at position 1.
> Random message -> Shown at position 2.
> Another $10 iOU request -> An expanded iOU preview message of $20 shown at position 3.
> Another random message -> Shown at position 4.
> New $30 request -> An expanded iOU preview message of $50 shown at position 5. The expanded iOU message at position 3 gets replaced by a simple iOU message of $10.
> 
> If the user pays the pending requests, the whole cycle stated above should reset.

### What is the root cause of that problem?

This is an improvement to an existing implementation

### What changes do you think we should make in order to solve the problem?

In order to implement the improvements in Chat Report, we can introduce a variable `latestIOUReportActionID` in IOU Report using which we can decide on show/hide of the IOU Action in Chat Report. The implementation can be as follows:

1) When there is a new IOU action added to the IOU Report, we will always set the `latestIOUReportActionID` to the `reportActionID` of the new IOU Action like this [here](https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/libs/actions/IOU.js#L559C1-L559C1):

`    iouReport.latestIOUReportActionID = iouAction.reportActionID;`

Further, the newly created IOU Action needs to be optimistically included in the Chat Report [here](https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/libs/actions/IOU.js#L192) like this. 

```
+                [iouAction.reportActionID]: iouAction,
                 [reportPreviewAction.reportActionID]: reportPreviewAction,
```

Also, `pendingAction` needs to be reset on success response for the IOU Action by adding the following [here](https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/libs/actions/IOU.js#L274):

                [iouAction.reportActionID]: {
                    pendingAction: null,
                },

2) To decide on whether to show/hide the IOU action within a Chat Report, we can check if `latestIOUReportActionID` is pointing to the current report action item. If so, we will not display this IOU Action since there is already a Report Preview that shows a consolidated preview of the IOU report. We can make the changes [here](https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/pages/home/report/ReportActionItem.js#L648) by adding the condition like this:

```

    // We do not want to show the IOU Action in Chat Report if this is the latest created one 
    if (
        props.action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU &&
        ReportUtils.isChatReport(props.report) && 
        props.iouReport && 
        props.iouReport.latestIOUReportActionID && 
        props.iouReport.latestIOUReportActionID === lodashGet(props.action, 'reportActionID', '0')
    ) {
        return null;
    }
```

Since the `iouReport` property is currently used only by ReportPreview action of Chat Report, we need to additionally support fetching for `iouReport` when the `actionName` is `iou`. We can do this by depending on the `IOUReportID` within `originalMessage` by replacing the [line here]() with this:

                const iouReportID = action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU ? action.originalMessage.IOUReportID : ReportActionsUtils.getIOUReportIDFromReportActionPreview(action);

3) Additionally, we need to handle the deletion of IOU Actions within Chat Report. We can do this in a similar way as it is done [here](https://github.com/Expensify/App/blob/d4a97bc7ab7d31e268daf9e7406e1edd0d773a6c/src/libs/actions/IOU.js#L2004C1-L2023C7). The other alternative is to reset the deleted IOU Action in Chat Report to null. However, this would mean that users will not be able to access the transaction thread conversation, if any. Maybe, we can take this up as an extended fix in PR as some discussion may be required. 

Additional Suggestion:
1) Once the IOU Report is settled, we may want to hide all the individual IOU Actions and show only the Report Preview. Considering the fact that the IOU Report is settled, it may be unnecessary information to show individual IOU Actions within the Chat Report. We may want to have a consensus here. Anyway, if we want to implement this, we can additionally check for `ReportUtils.isSettled()` within the condition of step (2) of the proposal.

BE Changes:
1) BE needs to support `latestIOUReportActionID`.
2) BE needs to include IOU Actions in Chat Reports.

### What alternative solutions did you explore? (Optional)

Alternative solution to using `latestIOUReportActionID` variable would be to compute run-time and find out the latest money request action for the non-settled IOUReport.

**Comment by Christinadobrzyn** [comment]
I think we're still reviewing proposals? @allroundexperts let me know if you'd like to see more!

**Comment by Christinadobrzyn** [comment]
@allroundexperts can you provide an update?

**Comment by Christinadobrzyn** [comment]
I'm going ooo until Monday Nov 27th so going to assign another BZ teammate to monitor this until I'm back. Thanks!

Status: We're reviewing proposals

**Comment by alexpensify** [comment]
I've reassigned the Bug label since I too am offline all next week. @MitchExpensify - tagging you in here to help keep this one moving forward. Thanks!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by MitchExpensify** [comment]
Cool, I got ya @alexpensify

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by MitchExpensify** [comment]
@allroundexperts what's the latest here?

**Comment by allroundexperts** [comment]
@MitchExpensify The behaviour has now changed. On staging its happening like below:
https://github.com/Expensify/App/assets/30054992/d13545a8-0bb7-41fc-aa0d-a0148c55e612

However, I think it should be like [this](https://github.com/Expensify/App/issues/30147#issuecomment-1799529847). If you agree then we're waiting for @roryabraham to respond [here](https://github.com/Expensify/App/issues/30147#issuecomment-1807696825).

Conversely, if you believe that the behaviour shown in above video is correct, then we should just close this and do nothing.

**Comment by MitchExpensify** [comment]
> However, I think it should be like https://github.com/Expensify/App/issues/30147#issuecomment-1799529847. If you agree then we're waiting for @roryabraham to respond https://github.com/Expensify/App/issues/30147#issuecomment-1807696825.

I do think the IOU should move to the most recent input in a chat when it gets a new request added to it. 

I'll defer to @Christinadobrzyn though as they'll be picking this back up from tomorrow!

**Comment by Christinadobrzyn** [comment]
Hi! I'm back, thanks for watching this @MitchExpensify! I reached out to Rory to see if we can get a review of this behaviour. 

https://github.com/Expensify/App/issues/30147#issuecomment-1799529847

**Comment by roryabraham** [comment]
> I do think the IOU should move to the most recent input in a chat when it gets a new request added to it.

I agree, but there should also be a record of the previous IOU request in the chat report, right?

**Comment by Christinadobrzyn** [comment]
hi! @allroundexperts how are these proposals looking? Is there anything I can do to help?

**Comment by allroundexperts** [comment]
@Christinadobrzyn Can we please get a confirmation for [this](https://github.com/Expensify/App/issues/30147#issuecomment-1830528143)?

**Comment by Christinadobrzyn** [comment]
Yeah I'll give it a try... I think we want this: 

The IOU to show in the chat at the instance it's sent in the chat thread:

![image](https://github.com/Expensify/App/assets/51066321/3bcce4fa-8c6d-4d6d-8592-1e77f23ab8df)

In addition, a summary at the bottom of the chat with the total amount due - this is the IOU that can be Paid:

![image](https://github.com/Expensify/App/assets/51066321/2b078aa2-b327-435e-82ef-6366130ffcc2)

Does this sound right @roryabraham?

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The money requests in-place are not persisted.

### What is the root cause of that problem?
This is more like a feature request, because the behavior of sending the plain-text reports was never implemented: in the chat report, only the IOU Preview is updated when placing the money request.

### What changes do you think we should make in order to solve the problem?
We should make several changes in this function:
https://github.com/Expensify/App/blob/385e8358ee1074b573fd5404b854e257284dda52/src/libs/actions/IOU.js#L116-L356

First, we need to optimistically add a new `reportAction` to the `chatReport` when placing a new money request. This should happen if there are already opened IOU requests – meaning the IOU Report is not going to be created. This action should be the last IOU action of the existing IOU Report.

It can be obtained by, for example:

```js
const allIouActions = ReportActionsUtils.getAllReportActions(iouReport.reportID);
const mostRecentIOURequestActionID = ReportActionsUtils.getMostRecentIOURequestActionID(Object.values(allIouActions));
const recentReportAction = ReportActionsUtils.getReportAction(iouReport.reportID, mostRecentIOURequestActionID);
```

Based on this actions' transaction data, we should build an optimistic plain-text report action and add it into the `chatReport` if `!isNewIOUReport` – meaning there is an active IOU request that was not published in the chat yet, because there was an IOU preview instead of it:

        {
            onyxMethod: isNewChatReport ? Onyx.METHOD.SET : Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${chatReport.reportID}`,
            value: {
                ...(isNewChatReport ? {[chatCreatedAction.reportActionID]: chatCreatedAction} : {}),
                ...(!isNewIOUReport ? {[lastIOURequestAction.reportActionID]: lastIOURequestAction} : {}),
                [reportPreviewAction.reportActionID]: reportPreviewAction,
            },
        },

https://github.com/Expensify/App/blob/385e8358ee1074b573fd5404b854e257284dda52/src/libs/actions/IOU.js#L159-L166C11

We'll also need to make BE changes so these actions get persisted on the BE and do not disappear on logout.

### What alternative solutions did you explore? (Optional)

**Comment by cooldev900** [comment]
[screen-capture (7).webm](https://github.com/Expensify/App/assets/13826499/2b788167-13ac-4844-9974-ea11bcb13bdf)

@Christinadobrzyn
Here is new result video.
Regarding recipt money request, what kind of text should be added in chat report?

**Comment by Christinadobrzyn** [comment]
Thanks @c3024! 

I think the text can just be a plain text of what the IOUs show. In your example "Camilo owes: ¥30" or "Camilo paid: ¥30" - what do you think @allroundexperts?

**Comment by roryabraham** [comment]
@Christinadobrzyn agreed for IOU payment actions. For previous requests I think it should be "You requested $20" or "Rory requested $20 (paid)"

**Comment by allroundexperts** [comment]
@cooldev900 Thanks for your proposal. Your proposal seems correct but I think @paultsimura's [proposal](https://github.com/Expensify/App/issues/30147#issuecomment-1843507650) is more simpler than yours and does not rely on storing / creating new IOU keys. As such, I think we should go with [there](https://github.com/Expensify/App/issues/30147#issuecomment-1843507650) proposal.


I would also agree with both of you that backend changes are needed for this so we add the plaintext version of the previous requests inside the chat report. @roryabraham Could you make those changes?

In the meantime, we can fix this for the optimistic case from the frontend. If any discrepancies come up after backend changes are in, we can deal with those in a followup issue.

@roryabraham Again, let us know if you agree with above.

🎀 👀 🎀 C+ reviewed

**Comment by grgia** [comment]
@roryabraham bump on https://github.com/Expensify/App/issues/30147#issuecomment-1849060470

@allroundexperts That said, I think if a solution requires backend changes at all, the issue should be made internal

**Comment by Christinadobrzyn** [comment]
ah sorry - I made this `internal` @allroundexperts let me know if you think it can remain External

**Comment by allroundexperts** [comment]
I'm not sure to be honest. I've seen a lot of issues that are external but require backend changes as well.

**Comment by Christinadobrzyn** [comment]
I think this could maybe be folded into our [VIP - Split P2P Chat Groups](https://github.com/Expensify/Expensify/milestone/796) so maybe we should keep this internal and then see if this could be resolved as part of that VIP. 

@GabiHExpensify could you review this GH and see if it's something that could be tied into the Split P2P VIP?

**Comment by Christinadobrzyn** [comment]
Asking if this fits into Wave 7 - https://expensify.slack.com/archives/C02NZ2HGJAZ/p1702937923040459

**Comment by grgia** [comment]
@allroundexperts What BE changes would be required for this in the case we went with your suggested proposal?

**Comment by allroundexperts** [comment]
@grgia currently, in a chat report, we have a single action of type `REPORTPREVIEW`. This action has a summary of all the IOU actions included in that IOU report. With the above proposals, we will need to store each individual item of the IOU report inside the chat report instead of just having a single item with a summary of the IOU report.

**Comment by youssef-lr** [comment]
This is not reproducible anymore. Also, note that in the recording of the issue the total number of requests goes from 7 to 6, that's when the REPORTPREVIEW moves up, then gets back to 7 and it moves down to the correct position.

https://github.com/Expensify/App/assets/9680864/1126dbeb-2b3f-4d75-be59-0d355c270523

Can we try reproducing this again please?

**Comment by roryabraham** [comment]
Sorry for the delay in responding. I know I [recently fixed](https://github.com/Expensify/Web-Expensify/pull/40008/files) and issue where the IOU preview jumps to the wrong place in the chat history.

It might be the case that a plaintext record of the original request is actually a new feature not a bug / scope creep, but it definitely makes sense to me and I think I recall seeing it in mockups way back when.

**Comment by roryabraham** [comment]
Slack discussion: https://expensify.slack.com/archives/C03U7DCU4/p1703104272565009

**Comment by roryabraham** [comment]
Ok, it sounds like the plaintext record of the IOU actions in the chat report is not currently expected / would be a new feature. Currently that stuff is tracked in the money request report (IOU/Expense report), but not in the chat report. The chat report only ever has 0 or 1 report preview actions, which move around in the history.

The original bug has been fixed here, and this is not a feature request to include more records in the chat report, so I'm going to close this.

**Comment by roryabraham** [comment]
For future context – we're eventually planning to have a record of IOU actions in the parent report (i.e: the chat report), but it's blocked on comment linking. When we have comment linking implemented, then we'll be able to post a link to IOU actions as a plaintext comment in the parent action (codename "post to parent" or "push to parent"). I think that's planned somewhere

**Comment by brunovjk** [comment]
Is this the same [issue](https://github.com/Expensify/App/issues/37245)?
